certora/funnels/Swapper.spec

```diff
@@ -0,0 +1,258 @@
+
+    setLimits(e, src, dst, cap, era);
+
+    mathint wardsAfter = wards(anyAddr);
+    mathint capSrcDstAfter; mathint eraSrcDstAfter; mathint dueSrcDstAfter; mathint endSrcDstAfter;
+    capSrcDstAfter, eraSrcDstAfter, dueSrcDstAfter, endSrcDstAfter = limits(src, dst);
+    mathint capOtherAfter; mathint eraOtherAfter; mathint dueOtherAfter; mathint endOtherAfter;
+    capOtherAfter, eraOtherAfter, dueOtherAfter, endOtherAfter = limits(otherAddr, otherAddr2);
+
+    assert wardsAfter == wardsBefore, "setLimits did not keep unchanged every wards[x]";
+    assert capSrcDstAfter == to_mathint(cap), "setLimits did not set limits[src][dst].cap to cap";
+    assert eraSrcDstAfter == to_mathint(era), "setLimits did not set limits[src][dst].era to era";
+    assert dueSrcDstAfter == 0, "setLimits did not set limits[src][dst].due to 0";
+    assert endSrcDstAfter == 0, "setLimits did not set limits[src][dst].end to 0";
+    assert capOtherAfter == capOtherBefore, "setLimits did not keep unchanged the rest of limits[x][y].cap";
+    assert eraOtherAfter == eraOtherBefore, "setLimits did not keep unchanged the rest of limits[x][y].era";
+    assert dueOtherAfter == dueOtherBefore, "setLimits did not keep unchanged the rest of limits[x][y].due";
+    assert endOtherAfter == endOtherBefore, "setLimits did not keep unchanged the rest of limits[x][y].end";
+}
+
+// Verify revert rules on setLimits
+rule setLimits_revert(address src, address dst, uint96 cap, uint32 era) {
+    env e;
+
+    bool canCall = roles.canCall(ilk(), e.msg.sender, currentContract, to_bytes4(0xf1b8ac2e));
+    mathint wardsSender = wards(e.msg.sender);
+
+    setLimits@withrevert(e, src, dst, cap, era);
+
+    bool revert1 = e.msg.value > 0;
+    bool revert2 = !canCall && wardsSender != 1;
+
+    assert revert1 => lastReverted, "revert1 failed";
+    assert revert2 => lastReverted, "revert2 failed";
+    assert lastReverted => revert1 || revert2, "Revert rules are not covering all the cases";
+}
+
+// Verify correct storage changes for non reverting swap
+rule swap(address src, address dst, uint256 amt, uint256 minOut, address callee, bytes data) {
+    env e;
+
+    require src == srcCon;
+    require dst == dstCon;
+    require callee == calleeCon;
+
+    address anyAddr;
+    address otherAddr;
+    address otherAddr2;
+    require otherAddr != src || otherAddr2 != dst;
+
+    require e.block.timestamp <= max_uint32;
+
+    address buffer = buffer();
+    require buffer != currentContract;
+    require buffer != callee;
+
+    mathint wardsBefore = wards(anyAddr);
+    mathint capBefore; mathint eraBefore; mathint dueBefore; mathint endBefore;
+    capBefore, eraBefore, dueBefore, endBefore = limits(src, dst);
+    mathint capOtherBefore; mathint eraOtherBefore; mathint dueOtherBefore; mathint endOtherBefore;
+    capOtherBefore, eraOtherBefore, dueOtherBefore, endOtherBefore = limits(otherAddr, otherAddr2);
+    mathint dstBalanceOfBufferBefore = dstCon.balanceOf(e, buffer);
+
```

Comment:
This rule really ought to be checking the change in `src` balance of the buffer as well. If I modify the Swapper code as follows, all rules pass:

```diff
...

interface GemLike {
...
+    function allowance(address, address) external returns (uint256);
}
...
    swap(...
        ...
 
-        GemLike(src).transferFrom(buffer, callee, amt);
+        require(GemLike(src).balanceOf(buffer) >= amt, "malicious compliance");
+        require(GemLike(src).allowance(buffer, address(this)) >= amt, "malicious compliance");
...
```

Granted I'm sure there are unit tests to catch this but it would make the spec a bit more complete.

Reply:
Done!